### PR TITLE
Update channel_session decorator to rehydrate http_session

### PR DESCRIPTION
**Pull request by linuxlewis**
This is my first take on a solution for #318 more feedback is welcomed.

I would like to add docs before this is merged because I think it might be a little tricky to understand. This is the use case as I understand it.

```python
@channel_session
@http_session
def ws_connect(message):
    pass

@channel_session
def ws_receive(message):
     print(message.http_session)
```

### CODE REVIEW

**Comment by andrewgodwin** [comment]
Yeah, this is roughly the approach I was thinking of, but it would be nice to:

* See if this works with both orderings of http_session/channel_session as decorators
* Consider making it an explicit argument to http_session if ordering silently makes it fail, and raise a loud error when it can't be done
* Add some docs

**Comment by linuxlewis** [comment]
@andrewgodwin I followed your suggested and added a explicit `persist=True` (better name ideas?) option to the `http_session` decorator to handle the cases when the channel_session isn't available (missing decorator, wrong order).

I've also updated the docstrings of both functions. I considered adding a paragraph in the getting-started page section on sessions but it felt out of place in context of the tutorial. Might be worth to create an issue for a separate sessions docs page to explain further. 

If there's a better place in the docs where I could add more detail point me in the direct and I'd be happy to.

File: channels/tests/test_sessions.py [hunk 1]
```diff
@@ -76,6 +76,27 @@ def inner(message):
...
+        """
+        message = Message({"reply_channel": "test-reply"}, None, None)
+
+        # Create a session with an http_session key
+        session = session_for_reply_channel("test-reply")
+        session[settings.SESSION_COOKIE_NAME] = session.session_key
+        session.save()
+
+        @channel_session
+        @channel_session
```
Comment (andrewgodwin) [inline channels/tests/test_sessions.py]: Why is this here twice?

**Comment by andrewgodwin** [comment]
Hrm, I'd rather the `persist` option also actually controlled if it tried to persist rather than just giving a different error - it's a bit odd that right now it persists even if you pass `persist=False`. Also left a comment on one of the test cases about the doubling of decorators.

I'm starting to wonder, though, if there's a cleaner overall approach here - what you've done works fine, but my original slightly-botched API idea for this stuff is not working super great with it. Any ideas how you might change the whole session persistence API (given decorators as the application mechanism and backwards-compat concerns)?

**Comment by linuxlewis** [comment]
I think the reason we're seeing it break down is because of the weird coordination between `http_session` and `channel_session` that needs to occur to make this work.  One idea for a cleaner API would be to have a separate decorator all together. It theory it would function similar to `channel_session_user_from_http`. I don't think it would conflict with any of the existing functionality either.

```python
def persistant_http_session(func):
  @http_session
  @channel_session
  def inner(message):
     #if write the session
     #else hydrate the session
  return inner
```

**Comment by andrewgodwin** [comment]
Yeah, that might end up being cleaner. Perhaps `channel_and_http_session` to keep with the naming scheme?

**Comment by linuxlewis** [comment]
This feels a lot cleaner. What are your thoughts?

**Comment by andrewgodwin** [comment]
Yes, I like that a lot more! I'll merge it in now - if you feel like putting in a PR for the docs too, that'd be great, else I'll get to it next week.
